src/libixy-vfio.c

```diff
@@ -0,0 +1,189 @@
+	char* group_name = basename(iommu_group_path);
+	int groupid;
+	ret = sscanf(group_name, "%d", &groupid);
+	if (ret != 1) {
+		error("Failed to convert group id '%s' to int", group_name);
+		return -1;
+	}
+
+	int firstsetup = 0; // Need to set up the container exactly once
+	int cfd = get_vfio_container();
+	if (cfd == -1) {
+		firstsetup = 1;
+		// open vfio file to create new cfio conainer
+		cfd = open("/dev/vfio/vfio", O_RDWR);
+		if (cfd < 0) {
+			error("Failed to open /dev/vfio/vfio");
+			return -1;
+		}
+		set_vfio_container(cfd);
+
+		// check if the container's API version is the same as the VFIO API's
+		if (ioctl(cfd, VFIO_GET_API_VERSION) != VFIO_API_VERSION) {
+			error("unknown VFIO API Version");
+			return -1;
+		}
+
+		// check if type1 is supported
+		if (ioctl(cfd, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU) != 1) {
+			error("container doesn't support Type1 IOMMU");
+			return -1;
+		}
+	}
+
+	// open VFIO group containing the device
+	snprintf(path, sizeof(path), "/dev/vfio/%d", groupid);
+	int vfio_gfd = open(path, O_RDWR);
+	if (vfio_gfd < 0) {
+		error("Failed to open vfio group");
+		return -1;
+	}
+
+	// check if group is viable
+	struct vfio_group_status group_status = {.argsz = sizeof(group_status)};
+	ret = ioctl(vfio_gfd, VFIO_GROUP_GET_STATUS, &group_status);
+	if (ret == -1) {
+		error("failed to get VFIO group status. errno: %d", errno);
+		return ret;
+	}
+	if (!group_status.flags & VFIO_GROUP_FLAGS_VIABLE) {
+		error("VFIO group is not viable - are all devices in the group bound "
+			  "to the VFIO driver?");
+		return -1;
+	}
+
+	// Add group to container
+	ret = ioctl(vfio_gfd, VFIO_GROUP_SET_CONTAINER, &cfd);
+	if (ret == -1) {
+		error("Failed to set container. errno: %d", errno);
+		return -1;
+	}
+
+	if (firstsetup != 0) {
+		// Set vfio type (type1 is for IOMMU like VT-d or AMD-Vi) for the
+		// container.
+		// This can only be done after at least one group is in the container.
+		ret = ioctl(cfd, VFIO_SET_IOMMU, VFIO_TYPE1_IOMMU);
+		if (ret != 0) {
+			error("Failed to set iommu type. errno: %d", -ret);
+			return -1;
+		}
+	}
+
+	// get device file descriptor
+	int vfio_fd = ioctl(vfio_gfd, VFIO_GROUP_GET_DEVICE_FD, pci_addr);
+	if (vfio_fd < 0) {
+		error("Cannot get device fd. errno: %d", -vfio_fd);
+		return -1;
+	}
+
+	// enable DMA
+	vfio_enable_dma(vfio_fd);
+
+	return vfio_fd;
+}
+
+// returns a uint8_t pointer to the MMAPED region or MAP_FAILED if failed
+uint8_t* vfio_map_region(int vfio_fd, int region_index) {
+	struct vfio_region_info region_info = {.argsz = sizeof(region_info)};
+	region_info.index = region_index;
+	int ret = ioctl(vfio_fd, VFIO_DEVICE_GET_REGION_INFO, &region_info);
+	if (ret == -1) {
+		// Failed to set iommu type
+		return MAP_FAILED; // MAP_FAILED == ((void *) -1)
+	}
+	return (uint8_t*) check_err(mmap(NULL, region_info.size,
+									 PROT_READ | PROT_WRITE, MAP_SHARED,
+									 vfio_fd, region_info.offset),
+								"mmap vfio bar0 resource");
```

Comment:
this formatting looks wrong with both tab size 4 and 8; I'd probably just merge it into one line?